lib/bats-core/test_functions.bash

```diff
@@ -3,30 +3,179 @@
 BATS_TEST_DIRNAME="${BATS_TEST_FILENAME%/*}"
 BATS_TEST_NAMES=()
 
-# Shorthand for source-ing files relative to the BATS_TEST_DIRNAME,
-# optionally with a .bash suffix appended. If the argument doesn't
-# resolve relative to BATS_TEST_DIRNAME it is sourced as-is.
-load() {
-  local file="${1:?}"
-
-  # For backwards-compatibility first look for a .bash-suffixed file.
-  # TODO consider flipping the order here; it would be more consistent
-  # and less surprising to look for an exact-match first.
-  if [[ -f "${BATS_TEST_DIRNAME}/${file}.bash" ]]; then
-    file="${BATS_TEST_DIRNAME}/${file}.bash"
-  elif [[ -f "${BATS_TEST_DIRNAME}/${file}" ]]; then
-    file="${BATS_TEST_DIRNAME}/${file}"
+# find_library_load_path expects an absolute path as the first argument
+# and echoes the first library load path for the bats library it finds.
+#
+# A library load path can be either a file or a directory.
+#
+# If library load path can not be found find_library_loader returns 1.
+#
+# The order for library loaders is:
+#   - file: <library_path>.bash
+#   - file: <library_path>
+#   - file: <library_path>/load.bash
+#   - file: <library_path>/load
+#   - directory: <library_path>
+find_library_load_path() {
+  local library_path="${1:?}"
+
+  local -a library_load_paths=(
+    "$library_path.bash"
+    "$library_path"
+    "$library_path/load.bash"
+    "$library_path/load"
+  )
+
+  for library_load_path in "${library_load_paths[@]}"; do
+    if [[ -f "$library_load_path" ]]; then
+      echo "$library_load_path"
+      return
+    fi
+  done
+
+  if [[ -d "$library_path" ]]; then
+    echo "$library_path"
+    return
   fi
 
-  if [[ ! -f "$file" ]] && ! type -P "$file" >/dev/null; then
-    printf 'bats: %s does not exist\n' "$file" >&2
-    exit 1
+  return 1
+}
+
+# find_in_bats_lib_path echoes the first recognized load path to
+# a library in BATS_LIB_PATH or relative to BATS_TEST_DIRNAME.
+#
+# Libraries relative to BATS_TEST_DIRNAME take precedence over
+# BATS_LIB_PATH.
+#
+# Library load paths are recognized using find_library_load_path.
+#
+# If no library is found find_in_bats_lib_path returns 1.
+find_in_bats_lib_path() {
+  local library_name="${1:?}"
+
+  local bats_lib_path="$BATS_LIB_PATH"
+  if [[ -z "$bats_lib_path" ]]; then
+    bats_lib_path="$HOME/.bats/lib:/usr/lib/bats"
```

Comment:
Maybe we should put that default into `libexec/bats-core/bats` instead and let users override later? This would allow them to prepend/append their own paths without removing the default paths, which is closer to the usual PATH handling.

Reply:
I decided to have no default value at all and to produce an error message when `bats_load_library` is used without setting `BATS_LIBRARY_PATH`. While the documentation encourages package maintainers to use `/usr/lib/bats` as the lib path, it might not be possible with package managers like `npm` or `brew`. This tries to strike a balance where we use the "natural flow" of the used package manager but unify where it is possible.